packages/terra-aggregator/src/Aggregator.jsx

```diff
@@ -0,0 +1,229 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const propTypes = {
+  /**
+   * The components that will be managed by the Aggregator. Each component provided will be provided with an 'aggregatorDelegate' prop
+   * containing the Aggregator API. Keys must also be provided to allow the Aggregator to uniquely identify the component.
+   * Immutable objects are preferred.
+   */
+  items: PropTypes.arrayOf(PropTypes.shape({
+    key: PropTypes.string.isRequired,
+    component: PropTypes.element.isRequired,
+  })),
+  /**
+   * A function that will override the Aggregators default render. The function will receive an Object parameter data necessary for the
+   * render process.
+   */
+  render: PropTypes.func,
+  /**
+   * A function that will be provided to Aggregator items that have received focus. The function must adhere to the standardized
+   * AppDelegate disclosure API.
+   */
+  disclose: PropTypes.func,
+};
+
+const defaultProps = {
+  items: [],
+};
+
+class Aggregator extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.requestFocus = this.requestFocus.bind(this);
+    this.releaseFocus = this.releaseFocus.bind(this);
+    this.setFocusState = this.setFocusState.bind(this);
+    this.resetFocusState = this.resetFocusState.bind(this);
+    this.renderItems = this.renderItems.bind(this);
+
+    this.state = {
+      focusedItemId: undefined,
+      focusedItemState: undefined,
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const { items } = this.props;
+    const { focusedItemId } = this.state;
+
+    if (nextProps.items !== this.props.items) {
+      // If the currently focused item is not present in the new items set,
+      // the focus is forcefully released to clean up any presented disclosures.
+      let focusItemIsPresent;
+      items.forEach((item) => {
+        if (item.key === focusedItemId) {
+          focusItemIsPresent = true;
+        }
+      });
+
+      if (!focusItemIsPresent) {
+        this.releaseFocus(undefined, true);
+      }
+    }
+  }
+
+  setFocusState(itemKey, itemState) {
+    this.setState({
+      focusedItemId: itemKey,
+      focusedItemState: itemState,
+    });
+  }
+
+  resetFocusState() {
+    this.setFocusState();
+  }
+
+  requestFocus(itemId, itemState) {
+    const { disclose } = this.props;
+    const { focusedItemId } = this.state;
+
+    return Promise.resolve()
+      .then(() =>
+        /**
+         * Focus is released on the currently focused item to ensure a clean start for the next component receiving focus.
+         * The releaseFocus's Promise is returned and inserted into the Promise chain to prevent disclosures from occurring
+         * if the focus release fails.
+         */
+        this.releaseFocus(focusedItemId)
+          .then(() => {
+            this.setFocusState(itemId, Object.freeze(itemState || {}));
```

Comment:
So you decided not to use redux for this?

Reply:
You're right, no redux. This is mostly a side-effect of the Promise-based logic in play; it gets very hard to replay the state.

Reply:
+1 on not using redux like you were before to persist the focus state, I could have seen some consumers abusing this.

Reply:
Yeah we're definitely going to have to come up with something. The Aggregator itself has few dirty data concerns. I moved a lot of that locking logic into the disclosure managers themselves; the Aggregator just hooks into the disclosure lifecycle and updates accordingly.  So we'll just have to figure out how to communicate disclosure locks to the larger overall application.

Reply:
Makes me wonder if the management of focus and dirty data should be separate concerns....

Reply:
Let me rephrase: the Aggregator has zero dirty data concerns. The only place where it could run into dirty data issues is when it tries to close an open disclosure during a focus release event, and the DisclosureManager encapsulates all of that. A component having focus is not 'dirty' state from the Aggregator's perspective. A component can make itself 'dirty' after having received focus, but that is the component's concern, not the Aggregator's.